File: counting_sort/C++/test/algorithm_test.cc
#include <gtest/gtest.h>
#include <c++14/make_unique.h>
#include <memory>
#include <vector>

#include "algorithm.h"

using namespace std;

namespace {

    // test sort

    TEST(AlgorithmTest, test_010_empty_array_is_sorted) {
        vector<int> source(0);
        vector<int> expect(0);
        unique_ptr<vector<int>> result = counting_sort(source);
        EXPECT_EQ(expect, *result);
    }

    TEST(AlgorithmTest, test_020_singleton_is_sorted) {
        vector<int> source {1};
        vector<int> expect {1};
        unique_ptr<vector<int>> result = counting_sort(source);
        EXPECT_EQ(1, source[0]);
        EXPECT_EQ(1, expect[0]);
        EXPECT_EQ(expect, *result);
    }

    TEST(AlgorithmTest, test_030_ordered_pair_is_sorted) {
        vector<int> source {1,2};
        vector<int> expect {1,2};
        unique_ptr<vector<int>> result = counting_sort(source);
        EXPECT_EQ(1, source[0]);
        EXPECT_EQ(1, expect[0]);
        EXPECT_EQ(expect, *result);
    }

    TEST(AlgorithmTest, test_040_reversed_pair_is_sorted) {
        vector<int> source {2,1};
        vector<int> expect {1,2};
        unique_ptr<vector<int>> result = counting_sort(source);
        EXPECT_EQ(2, source[0]);
        EXPECT_EQ(1, expect[0]);
        EXPECT_EQ(expect, *result);
    }

    TEST(AlgorithmTest, test_050_example_with_unique_ints_is_sorted) {
        vector<int> source {7, 1, 0, 9};
        vector<int> expect {0, 1, 7, 9};
        unique_ptr<vector<int>> result = counting_sort(source);
        EXPECT_EQ(expect, *result);
    }

    TEST(AlgorithmTest, test_060_example_with_repeats_is_sorted) {
        vector<int> source {4, 1, 3, 4, 3};
        vector<int> expect {1, 3, 3, 4, 4};
        unique_ptr<vector<int>> result = counting_sort(source);
        EXPECT_EQ(expect, *result);
    }

}

File: counting_sort/Python/test/algorithm_test.py
import pytest

from algorithm import counting_sort


# test sort

def test_empty_list_is_sorted():
    input, expect = [], []
    assert counting_sort(input) == expect


def test_singleton_is_sorted():
    input, expect = [1], [1]
    assert counting_sort(input) == expect


def test_ordered_pair_is_sorted():
    input, expect = [1, 2], [1, 2]
    assert counting_sort(input) == expect


def test_reversed_pair_is_sorted():
    input, expect = [2, 1], [1, 2]
    assert counting_sort(input) == expect


def test_example_with_unique_ints_is_sorted():
    input = [7, 1, 0, 9]
    expect = sorted(input)
    assert counting_sort(input) == expect


def test_example_with_repeats_is_sorted():
    input, expect = [4, 1, 3, 4, 3], [1, 3, 3, 4, 4]
    assert counting_sort(input) == expect
